junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/TestTemplateTestDescriptor.java

```diff
@@ -89,13 +90,18 @@ public JupiterEngineExecutionContext execute(JupiterEngineExecutionContext conte
 		ContainerExtensionContext containerExtensionContext = (ContainerExtensionContext) context.getExtensionContext();
 		List<TestTemplateInvocationContextProvider> providers = validateProviders(containerExtensionContext,
 			context.getExtensionRegistry());
-		AtomicInteger invocationIndex = new AtomicInteger();
-		// @formatter:off
-		providers.stream()
-				.flatMap(provider -> provider.provide(containerExtensionContext))
-				.map(invocationContext -> createInvocationTestDescriptor(invocationContext, invocationIndex.incrementAndGet()))
-				.forEach(invocationTestDescriptor -> execute(dynamicTestExecutor, invocationTestDescriptor));
-		// @formatter:on
+
+		int invocationIndex = 0;
+		for (TestTemplateInvocationContextProvider provider : providers) {
+			Stream<TestTemplateInvocationContext> invocationContextStream = provider.provide(containerExtensionContext);
+			Iterator<TestTemplateInvocationContext> contextIterator = invocationContextStream.iterator();
+			while (contextIterator.hasNext()) {
+				TestDescriptor invocationTestDescriptor = createInvocationTestDescriptor(contextIterator.next(),
+					++invocationIndex);
+				execute(dynamicTestExecutor, invocationTestDescriptor);
+			}
+			invocationContextStream.close();
```

Comment:
The stream should be used within a try-with-resources statement to ensure it's closed in case odf an exception.